compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -907,11 +907,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
             let ty = self.normalize(expr.span, ty);
             if self.can_coerce(found, ty) {
+                let mut node_id = self.tcx.hir().parent_id(expr.hir_id);
```

Comment:
Could you write an explanation for the logic here? I'm not sure I fully understand the stopping condition here.

Reply:
The intent is (I'm not 100% sure if this logic is correct) that we keep peeling away parents of the expression until we are no longer in a macro expansion (there might be multiple macro expansions in a chain, e.g. `outer!()` expands to `inner!()` etc.). Once we find a parent who is no longer from a macro expansion, we use that as a stopping condition for `find_ancestor_inside` to find a span that's contained within the span of the parent.

Reply:
If I don't do this "peeling away parents" business, it does not correctly work for `tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs` with just something like

```rust
let parent = self.tcx.hir().get_enclosing_scope(expr.hir_id).unwrap_or(expr.hir_id);
let parent_span = self.tcx.hir().span(parent);
let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
```

```
error[E0308]: mismatched types
 --> tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs:2:9
  |
2 |     Err(format!("error: {x}"))
  |         ^^^^^^^^^^^^^^^^^^^^^ expected `Box<dyn Error>`, found `String`
  |
  = note: expected struct `Box<dyn std::error::Error>`
             found struct `String`
  = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
 --> /home/jieyouxu/repos/rust/library/alloc/src/macros.rs:121:12
  |
12|         res.into()
  |            +++++++

error[E0308]: mismatched types
  --> tests/ui/typeck/issue-110017-format-into-help-deletes-macro.rs:20:9
   |
20 |     Err(outer!(x))
   |         ^^^^^^^^^ expected `Box<dyn Error>`, found `String`
   |
   = note: expected struct `Box<dyn std::error::Error>`
              found struct `String`
   = note: this error originates in the macro `format` which comes from the expansion of the macro `outer` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
  --> /home/jieyouxu/repos/rust/library/alloc/src/macros.rs:121:12
   |
121|         res.into()
   |            +++++++

```